aggregator/elem.go

```diff
@@ -0,0 +1,508 @@
+	// will be deleted once its aggregated values have been flushed
+	MarkAsTombstoned()
+
+	// Close closes the element
+	Close()
+}
+
+type elemBase struct {
+	sync.Mutex
+
+	opts       Options
+	id         metric.ID
+	policy     policy.Policy
+	tombstoned bool
+	closed     bool
+}
+
+type timedCounter struct {
+	timeNs  int64
+	counter aggregation.Counter
+}
+
+type timedTimer struct {
+	timeNs int64
+	timer  aggregation.Timer
+}
+
+type timedGauge struct {
+	timeNs int64
+	gauge  aggregation.Gauge
+}
+
+// CounterElem is the counter element
+type CounterElem struct {
+	elemBase
+
+	values []timedCounter // aggregated counters sorted by time in ascending order
+}
+
+// TimerElem is the timer element
+type TimerElem struct {
+	elemBase
+
+	values []timedTimer // aggregated timers sorted by time in ascending order
+}
+
+// GaugeElem is the gauge element
+type GaugeElem struct {
+	elemBase
+
+	values []timedGauge // aggregated gauges sorted by time in ascending order
+}
+
+// ResetSetData resets the counter and sets data
+func (e *elemBase) ResetSetData(id metric.ID, policy policy.Policy) {
+	e.id = id
+	e.policy = policy
+	e.tombstoned = false
+	e.closed = false
+}
+
+// MarkAsTombstoned marks an element as tombstoned, which means this element
+// will be deleted once its aggregated values have been flushed
+func (e *elemBase) MarkAsTombstoned() {
+	e.Lock()
+	if e.closed {
+		e.Unlock()
+		return
+	}
+	e.tombstoned = true
+	e.Unlock()
+}
+
+// NewCounterElem creates a new counter element
+func NewCounterElem(id metric.ID, policy policy.Policy, opts Options) *CounterElem {
+	return &CounterElem{
+		elemBase: elemBase{opts: opts, id: id, policy: policy},
+		values:   make([]timedCounter, 0, 2), // in most cases values will have two entries
+	}
+}
+
+// AddMetric adds a new counter value
+func (e *CounterElem) AddMetric(timestamp time.Time, mu unaggregated.MetricUnion) error {
+	alignedStart := timestamp.Truncate(e.policy.Resolution.Window).UnixNano()
+	e.Lock()
+	if e.closed {
+		e.Unlock()
+		return errCounterElemClosed
+	}
+	idx := e.findOrInsertWithLock(alignedStart)
+	e.values[idx].counter.Add(mu.CounterVal)
+	e.Unlock()
+	return nil
+}
+
+// ReadAndDiscard processes values before a given time and discards
+// them afterwards, returning whether the element can be collected
+// after discarding the values
+func (e *CounterElem) ReadAndDiscard(earlierThan time.Time, fn aggMetricFn) bool {
+	earlierThanNs := earlierThan.UnixNano()
+	e.Lock()
+	if e.closed {
+		e.Unlock()
+		return false
+	}
+	idx := 0
+	for range e.values {
+		// Bail as soon as the timestamp is no later than the target time
+		if e.values[idx].timeNs >= earlierThanNs {
+			break
+		}
+		endAt := time.Unix(0, e.values[idx].timeNs).Add(e.policy.Resolution.Window)
+		e.processValue(endAt, e.values[idx].counter, fn)
+		idx++
+	}
+	if idx > 0 {
+		// Shift remaining values to the left and shrink the values slice
+		n := copy(e.values[0:], e.values[idx:])
```

Comment:
Is this safe in terms of overlapping?  i.e. if this copy was right to left from dst to source then would be an issue.  If we're sure and its documented its left to right and won't clobber then obviously ok.